server/src/main/java/org/elasticsearch/search/rank/rrf/RRFRankContext.java

```diff
@@ -0,0 +1,162 @@
+        int queryCount = -1;
+        List<PriorityQueue<RRFRankDoc>> queues = new ArrayList<>();
+        for (QuerySearchResult querySearchResult : querySearchResults) {
+            if (querySearchResult.searchTimedOut()) {
+                topDocsStats.timedOut = true;
+                continue;
+            }
+            if (querySearchResult.terminatedEarly() != null && querySearchResult.terminatedEarly()) {
+                topDocsStats.terminatedEarly = true;
+            }
+            assert querySearchResult.getRankShardResult() instanceof RRFRankShardResult;
+            RRFRankShardResult rrfRankShardResult = (RRFRankShardResult) querySearchResult.getRankShardResult();
+
+            if (queryCount == -1) {
+                // we know we are on the first shard, so we create priority queues for each query
+                queryCount = rrfRankShardResult.queryCount;
+
+                for (int qi = 0; qi < queryCount; ++qi) {
+                    final int fqi = qi;
+                    queues.add(new PriorityQueue<>(windowSize + from) {
+                        @Override
+                        protected boolean lessThan(RRFRankDoc a, RRFRankDoc b) {
+                            float score1 = a.scores[fqi];
+                            float score2 = b.scores[fqi];
+                            if (score1 != score2) {
+                                return score1 < score2;
+                            }
+                            if (a.shardIndex != b.shardIndex) {
+                                return a.shardIndex > b.shardIndex;
+                            }
+                            return a.doc > b.doc;
+                        }
+                    });
+                }
+            }
+            assert queryCount == rrfRankShardResult.queryCount;
+
+            // for each query we add the appropriate docs based on their
+            // score for that query if they are part of the result set,
+            // skip otherwise
+            for (RRFRankDoc rrfRankDoc : rrfRankShardResult.rrfRankDocs) {
+                assert rrfRankDoc.shardIndex == -1;
+                rrfRankDoc.shardIndex = querySearchResult.getShardIndex();
+                for (int qi = 0; qi < queryCount; ++qi) {
+                    if (rrfRankDoc.positions[qi] != NO_RANK) {
+                        queues.get(qi).add(rrfRankDoc);
+                    }
+                }
+            }
+        }
+
+        // return early if we have no valid results
+        if (queues.isEmpty()) {
+            return SortedTopDocs.EMPTY;
+        }
+
+        // rank the global doc sets using RRF from the previously
+        // built priority queues
+        // the score is calculated on-the-fly where we update the
+        // score if we already saw it as part of a previous query's
+        // doc set, otherwise we make a new doc and calculate the
+        // initial score
+        Map<String, RRFRankDoc> results = new HashMap<>();
```

Comment:
Use a composite record class or something instead of concatenating strings.

https://richardstartin.github.io/posts/5-java-mundane-performance-tricks#use-wrappers-for-composite-hashmap-keys

Also, it would be good pre-size this hashmap if possible.

Reply:
Would you pre-size it to the max possible value every time?

Reply:
Implemented a record here for the key.

Reply:
> Would you pre-size it to the max possible value every time?

I would size it at least to something like the number of docs in the largest set of results. I don't think we want to dramatically OVERSIZE as a complete disjunction between the sets seems weird.